File: 23_nonAbundant.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

std::vector<int> getDivisors(int n) {
    std::vector<int> divisors;
    for (int i = 1; i < n; i++) {
        if (n % i == 0) {
            divisors.push_back(i);
        }
    }
    return divisors;
}

int main() {
    std::vector<int> abundant;
    const int LIMIT = 28123;

    for (int n = 1; n <= LIMIT; n++) {
        std::vector<int> divisors = getDivisors(n);
        int s = 0;
        for (int d : divisors) s += d;
        if (s > n) {
            abundant.push_back(n);
        }
    }

    std::unordered_set<int> abundantSums;
    for (int x : abundant) {
        for (int y : abundant) {
            if (x + y <= LIMIT) {
                abundantSums.insert(x + y);
            }
        }
    }

    long long result = 0;
    for (int i = 1; i <= LIMIT; i++) {
        if (abundantSums.find(i) == abundantSums.end()) {
            result += i;
        }
    }

    std::cout << result << std::endl;
    return 0;
}

File: 23_nonAbundant.py
def getDivisors(n):
    divisors = []

    for i in range(1, n):
        if n % i == 0:
            divisors.append(i)
    return divisors


def classify_numbers(limit):
    perfect = []
    abundant = []
    deficient = []

    for n in range(1, limit+1):
        divisors = getDivisors(n)
        s = sum(divisors)

        if s == n:
            perfect.append(n)

        elif s > n:
            abundant.append(n)

        elif s < n:
            deficient.append(n)
    
    return perfect, abundant, deficient

perfect, abundant, deficient = classify_numbers(28123)


abundant_sums = set()

for x in abundant:
    for y in abundant:
        if x + y <= 28123:
            abundant_sums.add(x + y)

result = 0

for i in range(1, 28124):
    if i not in abundant_sums:
        result += i

print(result)



